include/exadg/structure/spatial_discretization/operator.cpp

```diff
@@ -961,6 +961,8 @@ Operator<dim, Number>::solve_nonlinear(VectorType &       sol,
 
   //  elasticity_operator_nonlinear.set_inhomogeneous_boundary_values(sol);
 
+  affine_constraints_periodicity_and_hanging_nodes.distribute(sol); // <- these should be always enforced through matrix_free!?
```

Comment:
this should not be necessary, right? @nfehn 
When adding it, the visualization is correct *and* the iteration count decreses from 34.33 on average to 33.0.
So I think there is some inconsistency regarding the hangin gnode constraints, could that be?

Reply:
(these numbers are from the matrix-free case that behaves exactly the same as the matrix-based cased iteration-wise, when the diagonal entries in matrix-based are set to 1.0)

Reply:
The fact that the iteration counts change is strange, if this happens I believe we should look over the code again. The point is that `MatrixFree` should take care of the constraints whenever it needs it. Of course, we have the topic of how to apply the inhomogeneous/Dirichlet constraints discussed elsewhere, that needs a proper resolution to create a working code. But even that should not affect the iterations.

The `distribute` function is necessary when going to other functions that do not read out the solution vector via `FEEvaluation`, i.e., visualization/IO, specific functions, that do not resolve the constraints on the fly.

Reply:
Update: for the linear problems (unsteady + steady), the iterations are not changing (up to 2 decimal places) over 100 time steps for that problem:
![image](https://github.com/user-attachments/assets/7b47ae05-d388-4ade-ae2a-86e08c0696f7)
So it is a visualization problem for the linear case. For the nonlinear case, the linearization vector might be wrong (with the zero hanging node constraints enforced).

Reply:
I think we should at least keep the vector in a consistent state regarding constraints after each timestep. Ideally, the constraints would be in the vector directly after solving the system. I do not understand why this is not the case already ... even for linear problems.

Reply:
just as a reminder:
for a linear steady problem, commenting line
https://github.com/exadg/exadg/blob/6c8d7ca09cdf81dbfbd1f79980d38182e89225a2/include/exadg/structure/spatial_discretization/operator.cpp#L1010
gives:
![image](https://github.com/user-attachments/assets/c759b136-4c65-4e2c-8ca2-69915fb8201e)

where clearly, the boundary conditions are enforced (since that is the only thing nonzero), but the nodal values are zero constrained in the vector.